lib/base/io-engine.cpp

```diff
@@ -16,60 +16,83 @@
 
 using namespace icinga;
 
-CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc)
+CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc, boost::asio::io_context::strand& strand)
 	: m_Done(false)
 {
-	auto& ioEngine (IoEngine::Get());
+	VERIFY(strand.running_in_this_thread());
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+	auto& ie (IoEngine::Get());
+	Shared<AsioConditionVariable>::Ptr cv;
+
+	auto fastPath = [&ie] {
+		auto freeSlots (ie.m_CpuBoundSemaphore.load());
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+		while (freeSlots > 0) {
+			// A failure here is equivalent to loading the latest value into freeSlots
```

Comment:
A _failure_ sounds more serious than it actually is to me, thus I would rephrase this to something like: When the value of `m_CpuBoundSemaphore` was modified by some other threads after we loaded it earlier, the comparison in `compare_exchange_weak` will fail and write the newly loaded value into `freeSlots`.